src/main/java/Log4jHotPatch.java

```diff
@@ -50,123 +52,305 @@
   // property name for the agent version
   private static final String LOG4J_FIXER_AGENT_VERSION = "log4jFixerAgentVersion";
 
-  private static boolean verbose;
-
   private static boolean agentLoaded = false;
+  private static Logger logger;
+  private static Patcher patcher;
+  private static URLClassLoader patchLoader;
+
+  public interface Logger {
+    public void log(String msg);
+    public void setVerbose(boolean verbose);
+  }
+  public static class SimpleLogger implements Logger {
+    private static boolean verbose;
+    public SimpleLogger(boolean verbose) {
+      this.verbose = verbose;
+    }
+    @Override
+    public void setVerbose(boolean verbose) {
+      this.verbose = verbose;
+    }
+    @Override
+    public void log(String msg) {
+      if (verbose) {
+        System.out.println(msg);
+      }
+    }
+  }
 
-  private static boolean staticAgent = false; // Set to true if loaded as a static agent from 'premain()'
+  public interface Patcher {
+    public static final int SUCCESS = 0;
+    public static final int ERROR = 1;
+    public String getVersion();
+    public int install(String args, Instrumentation inst, Logger logger, boolean staticAgent);
+    public int uninstall();
+  }
 
-  private static void log(String message) {
-    if (verbose) {
-      System.out.println(message);
+  public static class EmptyPatcher implements Patcher {
+    private Logger logger;
+    public String getVersion() {
+      return "0";
+    }
+    @Override
+    public int install(String args, Instrumentation inst, Logger logger, boolean staticAgent) {
+      this.logger = logger;
+      logger.log("Installing Patcher version " + getVersion());
+      return SUCCESS;
+    }
+    @Override
+    public int uninstall() {
+      logger.log("Uninstalling Patcher version " + getVersion());
+      return SUCCESS;
     }
   }
 
-  public static void agentmain(String args, Instrumentation inst) {
+  public static class PatcherV1 implements Patcher {
+    private String args;
+    private Instrumentation inst;
+    private Logger logger;
+    private ClassFileTransformer transformer;
+    private static final String PATCHED_CLASS_INTERNAL = "org/apache/logging/log4j/core/lookup/JndiLookup";
+    private static final String PATCHED_CLASS_EXTERNAL = PATCHED_CLASS_INTERNAL.replace('/', '.');;
 
-    if (agentLoaded) {
-      log("Info: hot patch agent already loaded");
-      return;
+    @Override
+    public String getVersion() {
+      return "1";
```

Comment:
What do you think about setting this based on the classname so we don't have to change filename, classname and this value?

Reply:
You mean establish a naming convention for Patcher classes such that we could extract the version from the class name? E.g. something like classes have to end in "_<nr>", like `EmptyPach_0`, `Patcher_1`, etc..-

In general I like the idea except the fact that a "wrongly" named class would result in a runtime and not a compile-time error (unless we start using an annotation processor which is probably overkill).

What do others think?

Reply:
With tests we would find it at the PR stage and gradle runs the tests as part of the build. This issue isn't blocking the PR but would be nice to have.

Reply:
Sure why not :)